### ISSUE TITLE
Cannot Implement Multi-color Intertopic Distance Map Visualization

### ISSUE THREAD

**Issue by pariskang** [MaartenGr/BERTopic#1496]
I am trying to visualize topics using BERTopic with custom functions to support multi-color visualization in the intertopic distance map. Unfortunately, I failed it.

Here is the relevant code（_topics.py）:
import numpy as np
import pandas as pd
from umap import UMAP
from typing import List, Union
from sklearn.preprocessing import MinMaxScaler
import seaborn as sns

import plotly.express as px
import plotly.graph_objects as go

def visualize_topics(topic_model,
                     topics: List[int] = None,
                     top_n_topics: int = None,
                     custom_labels: Union[bool, str] = False,
                     title: str = "<b>Intertopic Distance Map</b>",
                     width: int = 650,
                     height: int = 650) -> go.Figure:
    """ Visualize topics, their sizes, and their corresponding words

    This visualization is highly inspired by LDAvis, a great visualization
    technique typically reserved for LDA.

    Arguments:
        topic_model: A fitted BERTopic instance.
        topics: A selection of topics to visualize
        top_n_topics: Only select the top n most frequent topics
        custom_labels: If bool, whether to use custom topic labels that were defined using 
                       `topic_model.set_topic_labels`.
                       If `str`, it uses labels from other aspects, e.g., "Aspect1".
        title: Title of the plot.
        width: The width of the figure.
        height: The height of the figure.

    Examples:

    To visualize the topics simply run:

    ```python
    topic_model.visualize_topics()
    ```

    Or if you want to save the resulting figure:

    ```python
    fig = topic_model.visualize_topics()
    fig.write_html("path/to/file.html")
    ```
    <iframe src="../../getting_started/visualization/viz.html"
    style="width:1000px; height: 680px; border: 0px;""></iframe>
    """

    # Select topics based on top_n and topics args
    freq_df = topic_model.get_topic_freq()
    freq_df = freq_df.loc[freq_df.Topic != -1, :]
    if topics is not None:
        topics = list(topics)
    elif top_n_topics is not None:
        topics = sorted(freq_df.Topic.to_list()[:top_n_topics])
    else:
        topics = sorted(freq_df.Topic.to_list())

    # Extract topic words and their frequencies
    topic_list = sorted(topics)
    frequencies = [topic_model.topic_sizes_[topic] for topic in topic_list]
    if isinstance(custom_labels, str):
        words = [[[str(topic), None]] + topic_model.topic_aspects_[custom_labels][topic] for topic in topic_list]
        words = ["_".join([label[0] for label in labels[:4]]) for labels in words]
        words = [label if len(label) < 30 else label[:27] + "..." for label in words]
    elif custom_labels and topic_model.custom_labels_ is not None:
        words = [topic_model.custom_labels_[topic + topic_model._outliers] for topic in topic_list]
    else:
        words = [" | ".join([word[0] for word in topic_model.get_topic(topic)[:5]]) for topic in topic_list]

##### 增加这个部分来生成一个颜色列表
    color_palette = sns.color_palette("hsv", len(topics)).as_hex()

    # Embed c-TF-IDF into 2D
    all_topics = sorted(list(topic_model.get_topics().keys()))
    indices = np.array([all_topics.index(topic) for topic in topics])

    if topic_model.topic_embeddings_ is not None:
        embeddings = topic_model.topic_embeddings_[indices]
        embeddings = UMAP(n_neighbors=2, n_components=2, metric='cosine', min_dist=0.3, random_state=42).fit_transform(embeddings)
    else:
        embeddings = topic_model.c_tf_idf_.toarray()[indices]
        embeddings = MinMaxScaler().fit_transform(embeddings)
        embeddings = UMAP(n_neighbors=2, n_components=2, metric='hellinger', min_dist=0.3, random_state=42).fit_transform(embeddings)

    df = pd.DataFrame({"x": embeddings[:, 0], "y": embeddings[:, 1],
                       "Topic": topic_list, "Words": words, "Size": frequencies, "Color": color_palette})

    return _plotly_topic_visualization(df, topic_list, title, width, height)

def _plotly_topic_visualization(df: pd.DataFrame,
                                topic_list: List[str],
                                title: str,
                                width: int,
                                height: int):
    """ Create plotly-based visualization of topics with a slider for topic selection """

    def get_color(topic_selected, num_topics):
        color_palette = sns.color_palette("hsv", num_topics).as_hex()  # 使用Seaborn库生成一个颜色列表
        if topic_selected == -1:
            marker_color = color_palette  # 如果没有选中任何主题，则使用整个颜色列表
        else:
            marker_color = ["red" if topic == topic_selected else color for topic, color in enumerate(color_palette)]
        return [{'marker.color': [marker_color]}]

    # Prepare figure range
    x_range = (df.x.min() - abs((df.x.min()) * .15), df.x.max() + abs((df.x.max()) * .15))
    y_range = (df.y.min() - abs((df.y.min()) * .15), df.y.max() + abs((df.y.max()) * .15))

    # Plot topics
    fig = px.scatter(df, x="x", y="y", size="Size", size_max=40, template="simple_white", labels={"x": "", "y": ""},
                     hover_data={"Topic": True, "Words": True, "Size": True, "x": False, "y": False}, color="Color")
    fig.update_traces(marker=dict(color="#B0BEC5", line=dict(width=2, color='DarkSlateGrey')))

    # Update hover order
    fig.update_traces(hovertemplate="<br>".join(["<b>Topic %{customdata[0]}</b>",
                                                 "%{customdata[1]}",
                                                 "Size: %{customdata[2]}"]))

    # Create a slider for topic selection
    #steps = [dict(label=f"Topic {topic}", method="update", args=get_color(topic)) for topic in topic_list]
    # 修正这一行，传入两个参数给 get_color
    steps = [dict(label=f"Topic {topic}", method="update", args=get_color(topic, len(topic_list))) for topic in topic_list]

    sliders = [dict(active=0, pad={"t": 50}, steps=steps)]

    # Stylize layout
    fig.update_layout(
        title={
            'text': f"{title}",
            'y': .95,
            'x': 0.5,
            'xanchor': 'center',
            'yanchor': 'top',
            'font': dict(
                size=22,
                color="Black")
        },
        width=width,
        height=height,
        hoverlabel=dict(
            bgcolor="white",
            font_size=16,
            font_family="Rockwell"
        ),
        xaxis={"visible": False},
        yaxis={"visible": False},
        sliders=sliders
    )

    # Update axes ranges
    fig.update_xaxes(range=x_range)
    fig.update_yaxes(range=y_range)

    # Add grid in a 'plus' shape
    fig.add_shape(type="line",
                  x0=sum(x_range) / 2, y0=y_range[0], x1=sum(x_range) / 2, y1=y_range[1],
                  line=dict(color="#CFD8DC", width=2))
    fig.add_shape(type="line",
                  x0=x_range[0], y0=sum(y_range) / 2, x1=x_range[1], y1=sum(y_range) / 2,
                  line=dict(color="#9E9E9E", width=2))
    fig.add_annotation(x=x_range[0], y=sum(y_range) / 2, text="D1", showarrow=False, yshift=10)
    fig.add_annotation(y=y_range[1], x=sum(x_range) / 2, text="D2", showarr
![newplot (22)](https://github.com/MaartenGr/BERTopic/assets/13728918/e9c8e62d-c79d-4eb6-aa11-4a8001fb2e5a)
ow=False, xshift=10)
    fig.data = fig.data[::-1]

    return fig

Thank you for your time and consideration. I look forward to your response.

Best regards, Paris Kang

**Comment by MaartenGr** [comment]
Thank you for opening up this issue. Unfortunately, the formatting of your code in Github is still a bit unclear. Could you make sure everything is formatted as Python? Also, could you go into more depth about what the issue is and what kind of resolution you are looking for? More sure to be complete in your description.

**Comment by pariskang** [comment]
Thank u for kindly and swiftly reply~I have updated the issue description just now

**Comment by MaartenGr** [comment]
You are giving it a color palette that is overwritten by the blue-ish color that you see here: 

https://github.com/MaartenGr/BERTopic/blob/951b97645acdf55e184889c761a83d2e1d73812f/bertopic/plotting/_topics.py#L91C5-L91C32

That function, `get_color`, defines the color that is chosen upon selection of the visualization. You would have to adjust the code there to show different colors. 

More specifically, I would advise removing the slider and therefore the `get_color` function as it will mess with your multi-color visualization.
